### PR TITLE
@pipe decorator

### PR DESCRIPTION
**Pull request by elijahbenizzy**
See https://github.com/DAGWorks-Inc/hamilton/issues/65, https://github.com/DAGWorks-Inc/hamilton/issues/372

## Changes
- [ ] Add pipe decorator

## How I tested this
- [ ] Lots of testing, end-to-end included

## Checklist

- [x] PR has an informative and human-readable title (this will be pulled into the release notes)
- [ ] Changes are limited to a single goal (no scope creep)
- [x] Code passed the pre-commit check & code is left cleaner/nicer than when first encountered.
- [x] Any _change_ in functionality is tested
- [x] New functions are documented (with a description, list of inputs, and expected output)
- [x] Placeholder code is flagged / future TODOs are captured in comments
- [x] Project documentation has been updated if adding/changing functionality.

### CODE REVIEW

File: docs/reference/decorators/with_columns.rst [hunk 1]
```diff
@@ -0,0 +1,12 @@
+=======================
+with_columns
+=======================
+
+** Overview **
+
+This is part of the hamilton pyspark integration. To install,
```
Comment (skrawcz) [inline docs/reference/decorators/with_columns.rst:7]: unfinished sentence.

File: hamilton/function_modifiers/macros.py [hunk 2]
```diff
@@ -289,3 +303,515 @@ def __init__(self, model_cls, config_param: str, **extra_model_params):
...
+        @pipe(
+            apply(_add_one),
+            apply(_multiply, y=2),
+            apply(_sum, y=value(3)),
+            apply(_multiply, y=source("upstream_node_to_multiply")),
+        )
+        def final_result(upstream_int: int) -> int:
+            return upstream_int
+
+    .. code-block:: python
```
Comment (skrawcz) [inline hamilton/function_modifiers/macros.py:571]: name this code.

File: hamilton/function_modifiers/macros.py [hunk 3]
```diff
@@ -289,3 +303,515 @@ def __init__(self, model_cls, config_param: str, **extra_model_params):
...
+    2. You want to pull in functions from an external repository, and build the DAG a little more procedurally
+    3. You want to use the same function multiple times, but with different parameters -- while `@does`/`@parameterize` can
+    do this, this presents an easier way to do this, especially in a chain.
+
+    To demonstrate the rules for chaining nodes, we'll be using the following example. This is
+    using primitives to demonstrate, but as hamilton is just functions of any python objects, this works perfectly with
+    dataframes, series, etc...
+
+
+    .. code-block:: python
```
Comment (skrawcz) [inline hamilton/function_modifiers/macros.py:551]: name this code. e.g. "example using @pipe & @apply... "

File: hamilton/function_modifiers/macros.py [hunk 4]
```diff
@@ -289,3 +303,515 @@ def __init__(self, model_cls, config_param: str, **extra_model_params):
...
+                _sum(
+                    _multiply(
+                        _add_one(upstream_int),
+                        y=2
+                    ),
+                    y=3
+                ),
+                y=upstream_node_to_multiply
+            )
+        )
```
Comment (skrawcz) [inline hamilton/function_modifiers/macros.py:586]: should show the other alternative of procedural code.

File: hamilton/function_modifiers/macros.py [hunk 5]
```diff
@@ -289,3 +303,515 @@ def __init__(self, model_cls, config_param: str, **extra_model_params):
...
+    where kwargs are some set of resolved upstream values. So, no `*args` are allowed, and `**kwargs` (variable keyword-only) are not
+    permitted. Note that this is not a design limitation, rather an implementation detail -- if you feel like you need this, please
+    reach out.
+
+    Furthermore, the function should be typed, as a Hamilton function would be.
+
+    One has two ways to tune the shape/implementation of the subsequent nodes:
+
+    1. `when`/`when_not`/`when_in`/`when_not_in` -- these are used to filter the application of the function. This is equivalent to calling
+        config...
```
Comment (skrawcz) [inline hamilton/function_modifiers/macros.py:601]: @config ...

File: hamilton/function_modifiers/macros.py [hunk 6]
```diff
@@ -289,3 +303,515 @@ def __init__(self, model_cls, config_param: str, **extra_model_params):
...
+    Note that functions must have no position-only arguments (this is rare in python, but hamilton does not handle these).
+    This basically means that the functions must be defined similarly to `def fn(x, y, z=10)` and not `def fn(x, y, /, z=10)`.
+    In fact, all arguments must be named and "kwarg-friendly", meaning that the function can happily be called with `**kwargs`,
+    where kwargs are some set of resolved upstream values. So, no `*args` are allowed, and `**kwargs` (variable keyword-only) are not
+    permitted. Note that this is not a design limitation, rather an implementation detail -- if you feel like you need this, please
+    reach out.
+
+    Furthermore, the function should be typed, as a Hamilton function would be.
+
+    One has two ways to tune the shape/implementation of the subsequent nodes:
```
Comment (skrawcz) [inline hamilton/function_modifiers/macros.py:598]: should say why.

_rather than if else statements within code and functions taking these configuration parameters_ 
...

File: hamilton/function_modifiers/macros.py [hunk 7]
```diff
@@ -289,3 +303,515 @@ def __init__(self, model_cls, config_param: str, **extra_model_params):
...
+        The second parameter is `namespace`. This is used to specify a namespace for the node. This is useful if you want
+        to either (a) ensure that the nodes are namespaced but share a common one to avoid name clashes (usual case), or (b)
+        if you want a custom namespace (unusual case). To indicate a custom namespace, one need simply pass in a string.
+
+        To indicate that a node should share a namespace with the rest of the apply(...) operations in a pipe, one can pass in `...` (the ellipsis).
+
+        .. code-block:: python
+
+            @pipe(
+                apply(_add_one).named("a", namespace="foo"), # foo.a
```
Comment (skrawcz) [inline hamilton/function_modifiers/macros.py:646]: alias I think is more common.

File: hamilton/function_modifiers/macros.py [hunk 8]
```diff
@@ -289,3 +303,515 @@ def __init__(self, model_cls, config_param: str, **extra_model_params):
...
+            @pipe(
+                apply(_add_one).when(foo="bar"),
+                apply(_add_two, y=source("other_node_to_add").when(foo="baz"),
+            )
+            def final_result(upstream_int: int) -> int:
+                return upstream_int
+
+        This will only apply the first function when the config parameter `foo` is set to `bar`, and the second when it is set to `baz`.
+
+    2. `named` -- this is used to name the node. This is useful if you want to refer to intermediate results. If this is left out,
```
Comment (skrawcz) [inline hamilton/function_modifiers/macros.py:616]: is `alias` better?

File: hamilton/function_modifiers/macros.py [hunk 9]
```diff
@@ -289,3 +303,515 @@ def __init__(self, model_cls, config_param: str, **extra_model_params):
...
+
+        if self.collapse:
+            raise NotImplementedError(
+                "Collapsing apply() functions as one node is not yet implemented for pipe(). Please reach out if you want this feature."
+            )
+
+        if self.chain:
+            raise NotImplementedError("@flow() is not yet supported -- this is ")
+
+    def inject_nodes(
```
Comment (skrawcz) [inline hamilton/function_modifiers/macros.py:694]: this function is too long, with too much logic? will be easier to test if it was smaller functions.

File: tests/resources/pipe.py [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+    return v
+
+
+def chain_1_not_using_pipe(v: int, input_1: int, input_2: int, calc_c: bool = False) -> int:
+    a = _add_one(v)
+    b = _add_two(a)
+    c = _add_n(b, n=3) if calc_c else b
+    d = _add_n(c, n=input_1)
+    e = _multiply_n(d, n=input_2)
+    return chain_1_using_pipe(e)
```
Comment (skrawcz) [inline tests/resources/pipe.py:45]: these are good examples to show one way versus the other.

File: hamilton/function_modifiers/macros.py (hunk 2 above)
Comment (elijahbenizzy) [reply hamilton/function_modifiers/macros.py:571]: Didn't know we can, we don't do that anywhere...

File: hamilton/function_modifiers/macros.py (hunk 9 above)
Comment (elijahbenizzy) [reply hamilton/function_modifiers/macros.py:694]: Meh it doesn't break nicely into smaller functions that need testing, and testing the smaller functions might be a little better for correctness, but would likely serve to lock down a contract that isn't what we want, and slow future development.

I'd rather have one larger function that has quite a few pieces that will change, with a clear contract and tests on it, then break it up purely for the sake of having higher test coverage.

Comment (skrawcz) [reply hamilton/function_modifiers/macros.py:713]: i dont' want to read it is the other feedback 😛

**Comment by skrawcz** [comment]
in person conversation notes:
 - apply --> step
 - stick with named
 - get example added

**Review by skrawcz — APPROVED** [review]

**Review by skrawcz — APPROVED** [review]
Looks good. Commits should have more design decision whys for your future self to refer back to.

**Pull request merged**